Implemented EZP-24237 Be sure to use dot as decimal separators for latitude and longitude

crevillo opened the pull request:
WIP
------

Implements https://jira.ez.no/browse/EZP-24237

See http://share.ez.no/forums/ez-publish-5-platform/maplocation-rendering-problem/

For reasons i still don't know, it happens that latitude and longitude are passed with comma as decimal separator to the MapView function 

This patch ensure we pass dot as decimal separator using 6 decimals.

crevillo commented:
ping @pspanja

pspanja commented on eZ/Bundle/EzPublishCoreBundle/Resources/views/content_fields.html.twig:
```diff
@@ -314,8 +314,8 @@
     {% set hasContent = field.value is not null %}
-    {% set latitude = field.value.latitude|default( 0 ) %}
-    {% set longitude = field.value.longitude|default( 0 ) %}
+    {% set latitude = field.value.latitude|number_format( 6, '.', '' )|default( 0 ) %}
```
`default` filter should be used before `number_format`. It seems to work like this as well, but that is because `number_format` will cast `null` value, so `default` will never be applied.

pspanja commented:
With order of filters fixed, I'm +1.

dpobel commented:
shouldn't this be done in the API somewhere instead ? from what you describe @crevillo , it seems like the latitude and longitude are string values instead of being floats ("45,5" instead of 45.5)

pspanja commented:
@dpobel that should not be the case, as the values are explicitly casted in external storage gateway: https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/FieldType/MapLocation/MapLocationStorage/Gateway/LegacyStorage.php#L233

@crevillo can you check that the `field.value.latitude` and `field.value.longitude` are indeed not strings in you case? 
With that, issue description should be updated to make clear that what this aims to fix is changing default number formatting as described in http://twig.sensiolabs.org/doc/filters/number_format.html.

andrerom commented:
Side note: Be aware that this is localization topic: http://en.wikipedia.org/wiki/Decimal_mark

crevillo commented:
Sorry guys, i'll be off until monday.  I'll check it thenes

El vie, 10 de abril de 2015 11:31, André R. <notifications@github.com>
escribió:

> Side note: Be aware that this is localization topic:
> http://en.wikipedia.org/wiki/Decimal_mark
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ezsystems/ezpublish-kernel/pull/1223#issuecomment-91492931>
> .
>

crevillo commented:
morning guys. As @pspanja says, latitude and longitute are explicity casted to float and as float they arrive to twig. 
Anyway, i've need to do some more testing to discard that this issue is not coming from my custom development. Actually what it bothers me is the following. In my pagelayout.html.twig template i've done some testing. i've added this just after my body tag
```twig
{% set n = 35.97 %}
{{ n }}
```
and no problem. the output is ```35.97``` with dot as decimal as expected. But if i move that piece of code some lines bellow, then the output turns to ```35,97``` with the comma. In the middle i have some includes and some render or render esi calls. 

i need to discard that something is doing something nasty in terms of changing localization or whatever. i would try to reproduce this in a demobundle installation so we all can check. 

Anyway, i've applied what Petar says and rebased this. This solve my issue but i'll try to find the root cause of this and i'll be back with the results.

crevillo commented:
@dpobel @pspanja i did a quick test with demobundle and that piece of code always show the dot as decimal separator. so actually this seems to be something from my custom development.
in one side, the number_format wouldn't hurt anybody but maybe it's not necessary at all. I edited my first comment to mark this as WIP. If you don't mind, i will kepp this pr open and i'll be back when i find the cause of my issue.

andrerom commented:
Should be ok for now, good luck :)

pspanja commented:
@crevillo 

> the number_format wouldn't hurt anybody but maybe it's not necessary at all

Probably, I think I initially made a wrong assumption when suggesting it should be fixed like that :)

crevillo commented:
@pspanja well, then we both did it wrong because that's what i thought too. and btw, that's what some people at symfony irc channels told me :)
anyway, what we actually know is either a twig thing or a bad code on my part (most probably).

crevillo commented:
@dpobel @pspanja @andrerom 
i've actually found something in my code. the site was a legacy entirely powered one and now is powered   by symfony stack almost entirely. anyway, stills needs legacy. what i found is twig numbers change decimal separator output to comma after a call to  ```path( 'ez_legacy' ... ```
i reproduce it having this pagelayout.html.twig
```twig
<!DOCTYPE html>
<head>

</head>
<body>
{% set n = 45.67 %}
{{ n }}
<a href="{{ path( 'ez_legacy', {'module_uri': '/content/search'} ) }}">Search</a>
{{ n }}
</body>
</html>
```
with this code i get ```45.67``` when the variable is first outputted and ```45,67``` 

The reason is in this old development i had (and don't ask me why) this thing in my frontend siteaccess
```
[RegionalSettings]
Locale=esl-ES
SystemLocale=es_ES,es_ES.utf8
```

Because of that, it seems that locale was set again in the [https://github.com/ezsystems/ezpublish-legacy/blob/master/kernel/private/classes/ezpkernelweb.php#L1182](ezpkernelweb class).

commeting that setting then i always get dots for decimal separators. 

Anyway, wondering if using diferent locales from the symfony perspective can lead to this thing too... i'll try to dig a bit more.

andrerom commented:
Ok, so maybe we need to consider injecting the locale set in Symfony then.
